[quantization] FP16 Dynamic Linear support

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30045 [quantization] FP16 Dynamic Linear support**

Depends on FBGEMM PR https://github.com/pytorch/FBGEMM/pull/201

Differential Revision: [D18580648](https://our.internmc.facebook.com/intern/diff/D18580648)

Commit:
[quantization] FP16 Linear support

[ghstack-poisoned]

jerryzh168 commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -66,10 +72,13 @@ def from_float(cls, mod):
+            weight_observer(mod.weight)
+            wt_scale, wt_zp = weight_observer.calculate_qparams()
+            qweight = torch.quantize_per_tensor(mod.weight.float(), float(wt_scale), int(wt_zp), torch.qint8)
+        else:
```
should we do check for `else if dtype == torch.float` here as well?

jerryzh168 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -48,10 +48,13 @@ def __init__(self, in_features, out_features, bias_=True):
+        if dtype == torch.qint8:
+            qweight = torch._empty_affine_quantized(
+                [out_features, in_features], scale=1, zero_point=0, dtype=torch.qint8)
+        else:
```
also here, check for `torch.float`?

Commit:
Update on "[quantization] FP16 Dynamic Linear support"


Depends on FBGEMM PR https://github.com/pytorch/FBGEMM/pull/201

Differential Revision: [D18580648](https://our.internmc.facebook.com/intern/diff/D18580648)

[ghstack-poisoned]

Commit:
Update on "[quantization] FP16 Dynamic Linear support"


Depends on FBGEMM PR https://github.com/pytorch/FBGEMM/pull/201

Differential Revision: [D18580648](https://our.internmc.facebook.com/intern/diff/D18580648)

[ghstack-poisoned]

Commit:
Update on "[quantization] FP16 Dynamic Linear support"


Depends on FBGEMM PR https://github.com/pytorch/FBGEMM/pull/201

Differential Revision: [D18580648](https://our.internmc.facebook.com/intern/diff/D18580648)

[ghstack-poisoned]

Commit:
Update on "[quantization] FP16 Dynamic Linear support"


Depends on FBGEMM PR https://github.com/pytorch/FBGEMM/pull/201

Differential Revision: [D18580648](https://our.internmc.facebook.com/intern/diff/D18580648)

[ghstack-poisoned]

jerryzh168 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -36,7 +36,7 @@ class Linear(torch.nn.Module):
     _FLOAT_MODULE = nn.Linear
 
-    def __init__(self, in_features, out_features, bias_=True):
+    def __init__(self, in_features, out_features, bias_=True, dtype=torch.qint8):
```
this seems to be changing the static quantization module, do we have static quantization kernel support for fp16 data?
Also can you add tests for this change and change the commit message to reflect this?

jamesr66a replied:
So long as the dynamic quantization module inherits from this module, this module needs to have the APIs necessary to support dynamic quantization, or we can make the dynamic module not inherit from this one.

As it is, `forward()` of the static module should fail if fp16 packed weights are given to it

jerryzh168 replied:
can we split the common part to another module then?

jianyuh commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -113,11 +120,17 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
 
     @torch.jit.export
     def __setstate__(self, state):
+        # This needs to go before self.set_weight_bias
```
Any reason?

jamesr66a replied:
Yes, otherwise `self.dtype` will be uninitialize and `set_weight_bias` cannot check that value

jianyuh replied:
Might be good to add that in the comments.

jianyuh commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -217,14 +217,79 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
+    Tensor output = at::empty(output_size, input.options().dtype(at::kFloat));
+
+    // Call the fp16 gemm interface
+    fbgemm::cblas_gemm_compute(
```
Do you also want to add the intra-op parallelization support in this PR? Up to you. I can add a follow-up PR for that soon.

jamesr66a replied:
I think that's better as a follow-up

jianyuh reviewed (commented):
Overall it looks great! Thanks for enabling this!

jamesr66a commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -27,17 +27,23 @@ class Linear(nnq.Linear):
         # Note that we can handle self.bias == None case.
-        Y = torch.ops.quantized.linear_dynamic(
-            x, self._packed_params)
+        if self.dtype == torch.qint8:
```
I think the existing convention is the qint types only, right?

jianyuh reviewed (approved):
LGTM! Thanks for completing this! Please make sure to address @jerryzh168 's comments.

jamesr66a commented:
@jianyuh actually this patch is not going to cleanly merge with master, so I'm going to have to do some work to rebase it sometime
